Sources/XCTest/Public/XCTestCase.swift

```diff
@@ -178,6 +179,29 @@ open class XCTestCase: XCTest {
     /// class.
     open class func tearDown() {}
 
+    private var teardownBlocks: [() -> Void] = []
+    private let teardownBlocksQueue: DispatchQueue = DispatchQueue(label: "org.swift.XCTest.XCTestCase.teardownBlocks")
+
+    /// Registers a block of teardown code to be run after the current test
+    /// method ends.
+    open func addTeardownBlock(_ block: @escaping () -> Void) {
+        teardownBlocksQueue.async {
+            self.teardownBlocks.append(block)
+        }
+    }
+
+    private func runTeardownBlocks() {
+        let blocks = teardownBlocksQueue.sync { () -> [() -> Void] in
+            let blocks = self.teardownBlocks
+            self.teardownBlocks = []
+            return blocks
+        }
+
```

Comment:
FYI, the Apple/ObjC implementation uses an internal flag to track whether or not the teardown blocks have been run yet or not, and raises an ObjC exception if a test attempts to add _more_ teardown blocks after they've already been run. This is to prevent weird cases like a test calling `addTeardownBlock(_:)` again inside a teardown block.

So you might consider matching that behavior here as well, using perhaps a `fatalError` instead of an exception.

Reply:
The last version includes this idea and my attempt at a test that proves the usage. The test is not perfect, because the precondition text doesn't seem to be printed even if the test is crashing.